dns-rc-autoscaler/nanny/nanny_lib.go

```diff
@@ -0,0 +1,90 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+/*
+Package nanny implements logic to poll the k8s apiserver for cluster status,
+and update a deployment based on that status.
+*/
+package nanny
+
+import (
+	"log"
+	"time"
+)
+
+// KubernetesClient is an object that performs the nanny's requisite interactions with Kubernetes.
+type KubernetesClient interface {
+	CountNodes() (uint64, uint64, error)
+	PodReplicas() (uint64, error)
+	UpdateReplicas(uint64) error
+	GetParentRc(namespace, podname string) (string, error)
+}
+
+// PollAPIServer periodically counts the number of nodes, estimates the expected
+// number of replicas, compares them to the actual replicas, and
+// updates the parent ReplicationController with the expected replicas if necessary.
+func PollAPIServer(k8s KubernetesClient, scaler Scaler, pollPeriod time.Duration, verbose bool) {
```

Comment:
Let me know when you're ready for another round.  Overall this is really
cool. I could see putting this in a whole new repo.  Or we could make one
repo for the various side-car apps we are building.

On Mon, Aug 1, 2016 at 10:19 PM, Tim Hockin <notifications@github.com>
wrote:

> In dns-rc-autoscaler/nanny/nanny_lib.go
> <https://github.com/kubernetes/contrib/pull/1427#discussion_r73095498>:
>
> > +	"log"
> > +	"time"
> > +)
> > +
> > +// KubernetesClient is an object that performs the nanny's requisite interactions with Kubernetes.
> > +type KubernetesClient interface {
> > +	CountNodes() (uint64, uint64, error)
> > +	PodReplicas() (uint64, error)
> > +	UpdateReplicas(uint64) error
> > +	GetParentRc(namespace, podname string) (string, error)
> > +}
> > +
> > +// PollAPIServer periodically counts the number of nodes, estimates the expected
> > +// number of replicas, compares them to the actual replicas, and
> > +// updates the parent ReplicationController with the expected replicas if necessary.
> > +func PollAPIServer(k8s KubernetesClient, scaler Scaler, pollPeriod time.Duration, verbose bool) {
>
> why poll? Watch should work for this, no? I guess poll is simpler, if less
> common.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/contrib/pull/1427/files/0bc392d4d1d2508aea02e27c3a8a6c34fe5e4a4f#r73095498>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVNo98oHkGO6oqyzFBmh4QBggnS47ks5qbtNvgaJpZM4JTRpr>
> .
>